api/renter.go

```diff
@@ -350,49 +387,156 @@ func (api *API) renterDeleteHandler(w http.ResponseWriter, req *http.Request, ps
 		return
 	}
 
-	siapath := strings.TrimPrefix(ps.ByName("siapath"), "/")
-	fileExists := false
+	// Set async flag to true.
+	req.URL.RawQuery += "&async=true"
+	p, errmsg := api.parseAndValidateDownloadParameters(w, req, ps)
+	if errmsg != nil {
+		WriteError(w, *errmsg, http.StatusBadRequest)
+		return
+	}
+
+	go api.renter.DownloadSection(p)
+
+	WriteSuccess(w)
+}
+
+func (api *API) parseAndValidateDownloadParameters(w http.ResponseWriter, req *http.Request, ps httprouter.Params) (*modules.RenterDownloadParameters, *Error) {
+	destination := req.FormValue("destination")
+
+	// The offset and length in bytes.
+	offsetparam := req.FormValue("offset")
+	lengthparam := req.FormValue("length")
+
+	// Determines whether the response is written to response body.
+	httprespparam := req.FormValue("httpresp")
 
-	files := api.renter.FileList()
-	for _, file := range files {
-		if file.SiaPath == siapath {
-			fileExists = true
+	// Determines whether to return on completion of download or straight away.
+	// If httprespparam is present, this parameter is ignored.
+	asyncparam := req.FormValue("async")
+
+	// Parse the offset and length parameters.
+	var offset, length uint64
+	var err error
+	if len(offsetparam) > 0 {
+		offset, err = strconv.ParseUint(offsetparam, 10, 64)
+		if err != nil {
+			return nil, &Error{"could not decode the offset as uint64: " +
+				err.Error()}
 		}
 	}
+	if len(lengthparam) > 0 {
+		length, err = strconv.ParseUint(lengthparam, 10, 64)
+		if err != nil {
+			return nil, &Error{"could not decode the length as uint64: " +
+				err.Error()}
+		}
+	}
+	// Verify that if either offset or length have been provided that both were provided.
+	offparampassed := len(offsetparam) > 0
+	lenparampassed := len(lengthparam) > 0
+	if (offparampassed || lenparampassed) &&
+		!(offparampassed && lenparampassed) {
+		var missingfield = "offset"
+		if lenparampassed {
+			missingfield = "length"
+		}
+		return nil, &Error{"either both \"offset\" and " +
+			"\"length\" have to be specified or neither. " +
+			missingfield + " has not been specified."}
+	}
 
-	if !fileExists {
-		WriteError(w, Error{"file at specified siapath does not exist"}, http.StatusBadRequest)
-		return
+	// Parse the httpresp parameter.
+	httpresp, errmsg := stringToBool(httprespparam)
+	if errmsg != nil {
+		return nil, errmsg
 	}
 
-	go api.renter.Download(strings.TrimPrefix(ps.ByName("siapath"), "/"), destination)
+	// Parse the async parameter.
+	async, errmsg := stringToBool(asyncparam)
+	if errmsg != nil {
+		return nil, errmsg
+	}
 
-	WriteSuccess(w)
+	if async && httpresp {
+		return nil, &Error{"only one of the async and httpresp flags can be specified."}
+	}
+
+	siapath := strings.TrimPrefix(ps.ByName("siapath"), "/") // Sia file name.
+
+	// Lookup the file associated with the nickname.
+	file, exists := api.renter.GetFile(siapath)
+	if !exists {
+		return nil, &Error{Message: "download failed: no file with that path"}
```

Comment:
probably worth printing which path was being requested. This will help us if the GUI for example is hitting this error and the user is showing us the error - we'll know better what's going on.